core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -413,6 +414,34 @@ public void writeMap(@Nullable Map<String, Object> map) throws IOException {
     }
 
     /**
+     * write map to stream with consistent order
+     * to make sure every map generated bytes order are same
+     */
+    public void writeMapWithConsistentOrder(@Nullable Map<String, Object> map) throws IOException {
+        if (map == null) {
+            writeByte((byte) -1);
+            return;
+        }
+        if (map instanceof LinkedHashMap) {
+            this.writeByte((byte) 9);
+        } else {
+            this.writeByte((byte) 10);
+        }
+        this.writeVInt(map.size());
+        Iterator<Map.Entry<String, Object>> iterator =
+            map.entrySet().stream().sorted((a, b) -> a.getKey().compareTo(b.getKey())).iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, Object> next = iterator.next();
+            this.writeString(next.getKey());
+            if (next.getValue() instanceof Map) {
+                writeMapWithConsistentOrder((Map<String, Object>) next.getValue());
```

Comment:
I think you should just always use `writeGenericValue` for the value. Unless it doesn't work for something?